// zksync-ethers/build-esm/typechain/factories/L2InteropRootStorage__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, Interface } from "ethers";
const _abi = [
    {
        type: "function",
        name: "addInteropRoot",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "blockOrBatchNumber",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "sides",
                type: "bytes32[]",
                internalType: "bytes32[]",
            },
        ],
        outputs: [],
        stateMutability: "nonpayable",
    },
    {
        type: "function",
        name: "interopRoots",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "blockOrBatchNumber",
                type: "uint256",
                internalType: "uint256",
            },
        ],
        outputs: [
            {
                name: "interopRoot",
                type: "bytes32",
                internalType: "bytes32",
            },
        ],
        stateMutability: "view",
    },
    {
        type: "event",
        name: "InteropRootAdded",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                indexed: true,
                internalType: "uint256",
            },
            {
                name: "blockNumber",
                type: "uint256",
                indexed: true,
                internalType: "uint256",
            },
            {
                name: "sides",
                type: "bytes32[]",
                indexed: false,
                internalType: "bytes32[]",
            },
        ],
        anonymous: false,
    },
    {
        type: "error",
        name: "CallerMustBeBootloader",
        inputs: [],
    },
    {
        type: "error",
        name: "InteropRootAlreadyExists",
        inputs: [],
    },
    {
        type: "error",
        name: "MessageRootIsZero",
        inputs: [],
    },
    {
        type: "error",
        name: "SidesLengthNotOne",
        inputs: [],
    },
];
export class L2InteropRootStorage__factory {
    static createInterface() {
        return new Interface(_abi);
    }
    static connect(address, runner) {
        return new Contract(address, _abi, runner);
    }
}
L2InteropRootStorage__factory.abi = _abi;
