// zksync-ethers/src/typechain/factories/L2InteropRootStorage__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  L2InteropRootStorage,
  L2InteropRootStorageInterface,
} from "../L2InteropRootStorage";

const _abi = [
  {
    type: "function",
    name: "addInteropRoot",
    inputs: [
      {
        name: "chainId",
        type: "uint256",
        internalType: "uint256",
      },
      {
        name: "blockOrBatchNumber",
        type: "uint256",
        internalType: "uint256",
      },
      {
        name: "sides",
        type: "bytes32[]",
        internalType: "bytes32[]",
      },
    ],
    outputs: [],
    stateMutability: "nonpayable",
  },
  {
    type: "function",
    name: "interopRoots",
    inputs: [
      {
        name: "chainId",
        type: "uint256",
        internalType: "uint256",
      },
      {
        name: "blockOrBatchNumber",
        type: "uint256",
        internalType: "uint256",
      },
    ],
    outputs: [
      {
        name: "interopRoot",
        type: "bytes32",
        internalType: "bytes32",
      },
    ],
    stateMutability: "view",
  },
  {
    type: "event",
    name: "InteropRootAdded",
    inputs: [
      {
        name: "chainId",
        type: "uint256",
        indexed: true,
        internalType: "uint256",
      },
      {
        name: "blockNumber",
        type: "uint256",
        indexed: true,
        internalType: "uint256",
      },
      {
        name: "sides",
        type: "bytes32[]",
        indexed: false,
        internalType: "bytes32[]",
      },
    ],
    anonymous: false,
  },
  {
    type: "error",
    name: "CallerMustBeBootloader",
    inputs: [],
  },
  {
    type: "error",
    name: "InteropRootAlreadyExists",
    inputs: [],
  },
  {
    type: "error",
    name: "MessageRootIsZero",
    inputs: [],
  },
  {
    type: "error",
    name: "SidesLengthNotOne",
    inputs: [],
  },
] as const;

export class L2InteropRootStorage__factory {
  static readonly abi = _abi;
  static createInterface(): L2InteropRootStorageInterface {
    return new Interface(_abi) as L2InteropRootStorageInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): L2InteropRootStorage {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as L2InteropRootStorage;
  }
}
